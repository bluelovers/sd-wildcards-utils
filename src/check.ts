import { Document, isDocument, isNode, Node } from 'yaml';
import {
	IFindPathEntry,
	IRecordWildcards,
	IOptionsCheckAllSelfLinkWildcardsExists,
} from './types';
import {
	assertWildcardsPath,
	convertWildcardsPathsToName,
	parseWildcardsYaml,
} from './index';
import {
	convertWildcardsNameToPaths,
	matchDynamicPromptsWildcardsAll,
} from './util';
import {
	findPath
} from './node-find';
import picomatch, { Matcher } from 'picomatch';

/**
 * Checks if all self-link wildcards exist in a given object.
 *
 * @param obj - The object to check, can be a YAML string, Uint8Array, or a YAML Document/Node.
 * @param chkOpts - Optional options for the check.
 * @returns An object containing the results of the check.
 *
 * @throws Will throw an error if the provided object is not a YAML Document/Node and cannot be parsed as a YAML string.
 *
 * @remarks
 * This function will parse the provided object into a YAML Document/Node if it is not already one.
 * It will then extract all self-link wildcards from the YAML string representation of the object.
 * For each wildcard, it will check if it exists in the JSON representation of the object using the `findPath` function.
 * The function will return an object containing arrays of wildcard names that exist, do not exist, or were ignored due to the ignore option.
 * It will also include an array of any errors that occurred during the check.
 */
export function checkAllSelfLinkWildcardsExists(obj: IRecordWildcards | Node | Document | string | Uint8Array, chkOpts?: IOptionsCheckAllSelfLinkWildcardsExists)
{
	chkOpts ??= {};

	const maxErrors = chkOpts.maxErrors > 0 ? chkOpts.maxErrors : 10;

	if (!(isDocument(obj) || isNode(obj)))
	{
		obj = parseWildcardsYaml(obj as string)
	}

	const str = obj.toString();
	const json = obj.toJSON();

 let entries = matchDynamicPromptsWildcardsAll(str, {
	 unsafe: true,
	 ...chkOpts.optsMatch,
	 unique: true,
 });

	let isMatchIgnore: Matcher = () => false;

	if (chkOpts.ignore?.length)
	{
		isMatchIgnore = picomatch(chkOpts.ignore);
	}

	const listHasExists: string[] = [];
	const listHasExistsWildcards: string[] = [];
	const ignoreList: string[] = [];

	const errors: Error[] = [];

	for (const entry of entries)
	{
		if (isMatchIgnore(entry.name))
		{
			ignoreList.push(entry.name);
			continue;
		}

		const paths = convertWildcardsNameToPaths(entry.name);

		// @ts-ignore
		let list: IFindPathEntry[] = [];

		try
		{
			assertWildcardsPath(entry.name);

			list = findPath(json, paths, {
				onlyFirstMatchAll: true,
				throwWhenNotFound: true,
				allowWildcardsAtEndMatchRecord: chkOpts.allowWildcardsAtEndMatchRecord,
			});

			if (chkOpts.report)
			{
				listHasExists.push(...list.map(v => convertWildcardsPathsToName(v.key)));

				if (entry.name.includes('*'))
				{
					listHasExistsWildcards.push(entry.name);
				}
			}
		}
		catch (e)
		{
			errors.push(e as any)

			if (errors.length >= maxErrors)
			{
				let e2 = new RangeError(`Max Errors. errors.length ${errors.length} >= ${maxErrors}`);
				errors.unshift(e2)

				break;
			}

			continue;
		}
	}

	return {
		obj,
		listHasExists,
		listHasExistsWildcards,
		ignoreList,
		errors,
	}
}

