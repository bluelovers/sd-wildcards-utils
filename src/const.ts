export const RE_UNSAFE_QUOTE = /['"]/;
export const RE_UNSAFE_QUOTE_DOUBLE = /"/;
export const RE_UNSAFE_VALUE = /^\s*-|[{$~!@}\n|:?#'"%]/;
//export const RE_UNSAFE_PLAIN = /-/;
export const RE_UNSAFE_PLAIN = /-/;
