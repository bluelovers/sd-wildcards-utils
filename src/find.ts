
import { isMatch } from 'picomatch';
import {
	ICachesFindPath,
	IFindPathEntry,
	IOptionsFind,
	IRecordWildcards,
	IVisitPathsListReadonly,
	IWildcardsYAMLDocument,
} from './types';
import { Document, isDocument } from 'yaml';
import { PicomatchOptions } from 'picomatch';
import { convertWildcardsPathsToName } from './util';

export function pathsToWildcardsPath(paths: IVisitPathsListReadonly, full?: boolean)
{
	let s = convertWildcardsPathsToName(paths);
	if (full)
	{
		s = `__${s}__`
	}
	return s
}

export function pathsToDotPath(paths: IVisitPathsListReadonly)
{
	return paths.join('.');
}

/**
 * Recursively searches for a path in a nested object or array structure.
 */
export function findPath(data: IRecordWildcards | Document | IWildcardsYAMLDocument,
	paths: string[],
	findOpts?: IOptionsFind,
	prefix: string[] = [],
	list: IFindPathEntry[] = []
)
{
	findOpts ??= {};
	prefix ??= [];
	list ??= [];

	let _cache: ICachesFindPath = {
		paths: paths.slice(),
		findOpts,
		prefix,
		globOpts: findPathOptionsToGlobOptions(findOpts),
	}

	if (isDocument(data))
	{
		// @ts-ignore
		_cache.data = data;

		data = data.toJSON() as IRecordWildcards;
	}

	return _findPathCore(data, paths.slice(), findOpts, prefix, list, _cache)
}

export function findPathOptionsToGlobOptions(findOpts?: IOptionsFind): PicomatchOptions
{
	return {
		...findOpts?.globOpts,
		ignore: findOpts?.ignore,
	} satisfies PicomatchOptions
}

export function _findPathCore(data: IRecordWildcards, paths: string[], findOpts: IOptionsFind, prefix: string[], list: IFindPathEntry[], _cache: ICachesFindPath)
{
	paths = paths.slice(); // Create a copy of the paths array to avoid modifying the original array.
	const current = paths.shift(); // Remove the first element from the paths array.
	const deep = paths.length > 0; // Check if there are remaining paths to search.

	for (const key in data)
	{
		if (findOpts.onlyFirstMatchAll && list.length)
		{
			break;
		}

		// Create the current path.
		const target = prefix.slice().concat(key);
		const search = prefix.slice().concat(current);

		// Check if the current key matches the current path element.
		//const bool = isMatch(key, current);
		const bool = isMatch(pathsToWildcardsPath(target), pathsToWildcardsPath(search), _cache.globOpts);

		if (bool)
		{

			const value = data[key]; // Get the value at the current path.

			const notArray = !Array.isArray(value); // Check if the value is not an array.

			if (deep)
			{
				if (notArray && typeof value !== 'string')
				{
					_findPathCore(value, paths, findOpts, target, list, _cache); // Recursively search for the remaining paths in the nested object or array.
					continue;
				}
			}
			else if (!notArray)
			{
				list.push({
					key: target,
					value,
				}); // Add the found path and its corresponding value to the list.
				continue;
			}
			else if (!deep && _cache.findOpts.allowWildcardsAtEndMatchRecord && current.includes('*') && typeof value === 'object' && value)
			{
				list.push({
					key: target,
					value,
				}); // Add the found path and its corresponding value to the list.
				continue;
			}

			if (!current.includes('*') || notArray && !deep)
			{
				throw new TypeError(`Invalid Type. paths: [${target}], isMatch: ${bool}, deep: ${deep}, deep paths: [${paths}], notArray: ${notArray}, match: [${search}], value: ${value}, _cache : ${JSON.stringify(_cache)}`); // Throw an error if the value is not a string and there are remaining paths to search.
			}
		}
	}

	if (prefix.length === 0 && findOpts.throwWhenNotFound && !list.length)
	{
		throw new RangeError(`Invalid Paths. paths: [${[current, ...paths]}], _cache : ${JSON.stringify(_cache)}`);
	}

	return list; // Return the list of found paths and their corresponding values.
}
