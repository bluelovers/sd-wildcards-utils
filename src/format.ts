import { IOptionsSharedWildcardsYaml } from './types';

export function stripZeroStr(value: string)
{
	return value
		.replace(/[\x00\u200b]+/g, '')
}

export function trimPrompts(value: string)
{
	return value
		.replace(/\xa0/g, ' ')
		.replace(/^\s+|\s+$/g, '')
		.replace(/^\s+|\s+$/gm, '')
		.replace(/\n\s*\n/g, '\n')
		.replace(/\s+/gm, ' ')
		.replace(/[ ,.]+(?=,|$)/gm, '')
		.replace(/,\s*(?=,|$)/g, '')
		;
}

export function normalizeWildcardsYamlString(value: string)
{
	value = stripZeroStr(value)
		.replace(/\xa0/g, ' ')
		.replace(/[,.]+(?=,)/gm, '')
		.replace(/[ .]+$/gm, '')
		.replace(/\{\s+(\d+(?:\.\d+)?(?:-(?:\d+(?:\.\d+)?)?\$\$|::))/gm, '{$1')
		.replace(/\|\s(\d+(?:\.\d+)?::)/gm, '|$1')
		.replace(/^[ \t]+-[ \t]*$/gm, '')
		.replace(/^([ \t]+-)[ \t][ ,.]+/gm, '$1 ')
		.replace(/^([ \t]+-[^\n]+),+$/gm, '$1')
	;
	return value
}

export function formatPrompts(value: string, opts?: IOptionsSharedWildcardsYaml)
{
	opts ??= {};

	value = stripZeroStr(value);
	value = trimPrompts(value);
	value = normalizeWildcardsYamlString(value);

	if (opts.minifyPrompts)
	{
		value = value
			.replace(/(,)\s+/gm, '$1')
			.replace(/\s+(,)/gm, '$1')
			.replace(/(?<=,\|})\s+/gm, '')
			.replace(/\s+(?=\{(?:\s*\d+(?:\.\d+)?::)?,)/gm, '')
		;
	}

	return value
}

export function stripBlankLines(value: string, appendEOF?: boolean)
{
	value = value
		.replace(/(\r?\n)[\s\r\n\t\xa0]+(\r?\n)/g, '$1$2')
		.replace(/(\r?\n)(?:\r?\n)(?=[\s\t\xa0])/g, '$1')
		.replace(/[ \xa0\t]+$/gm, '')
	;

	if (appendEOF)
	{
		value = value.replace(/\s+$/, '');
		value += '\n\n';
	}

	return value
}
