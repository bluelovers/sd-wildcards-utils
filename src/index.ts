/**
 * Created by user on 2024/5/21.
 */
import { Document, isDocument, ParsedNode, parseDocument, stringify, YAMLMap } from 'yaml';
import {
	defaultOptionsParseDocument,
	defaultOptionsStringify,
	getOptionsFromDocument,
} from './options';
import { _visitNormalizeScalar, visitWildcardsYAML } from './node/node-items';
import { createDefaultVisitWildcardsYAMLOptions, validWildcardsYamlData } from './valid';
import {
	IOptionsParseDocument,
	IOptionsStringify,
	IOptionsVisitor,
	IParseWildcardsYamlInputSource,
	IRecordWildcards,
	IWildcardsYAMLDocument,
	IWildcardsYAMLDocumentParsed,
	IWildcardsYAMLMapRoot,
} from './types';
import { _expandForwardSlashKeys } from './parser';

export * from './util';
export * from './options';
export * from './node/node-items';
export * from './valid';
export * from './prompts/valid-prompts';
export * from './node/node-merge';
export * from './node/node-find';
export * from './prompts/format';
export * from './prompts/prompts';
export * from './check';
export * from './node/node-is';
export type * from './types';

/**
 * Normalizes a YAML document by applying specific rules to its nodes.
 **/
export function normalizeDocument<T extends Document>(doc: T, opts?: IOptionsParseDocument)
{
	let options = getOptionsFromDocument(doc, opts);

	const defaults = createDefaultVisitWildcardsYAMLOptions(options);

	let checkUnsafeQuote = !options.disableUnsafeQuote;

	let visitorOptions: IOptionsVisitor = {
		...defaults,

		Scalar(key, node, parentNodes)
		{
			return _visitNormalizeScalar(key, node, parentNodes, {
				checkUnsafeQuote,
				options,
			})
		},
	};

	visitWildcardsYAML(doc, visitorOptions)
}

/**
 * Converts the given YAML data to a string, applying normalization and formatting.
 *
 * @returns - A string representation of the input YAML data, with normalization and formatting applied.
 *
 * @throws - Throws a `SyntaxError` if the input data is invalid according to the `validWildcardsYamlData` function.
 *
 * @remarks
 * This function takes the input YAML data and applies normalization and formatting using the provided options.
 * If the input data is a `Document` object, it first normalizes the document using the `normalizeDocument` function.
 * Then, it converts the normalized document to a string using the `toString` method with the specified options.
 * If the input data is not a `Document` object, it directly converts the data to a string using the `stringify` function with the specified options.
 *
 * @example
 * ```typescript
 * const yamlData: IRecordWildcards = {
 *   key1: ['value1', 'value2'],
 *   key2: {
 *     subkey1: ['value3', 'value4'],
 *   },
 * };
 *
 * const yamlString = stringifyWildcardsYamlData(yamlData);
 * console.log(yamlString);
 * // Output:
 * // key1:
 * //   - value1
 * //   - value2
 * // key2:
 * //   subkey1:
 * //     - value3
 * //     - value4
 * ```
 */
export function stringifyWildcardsYamlData<T extends IRecordWildcards | IWildcardsYAMLDocument | Document>(data: T | unknown,
	opts?: IOptionsStringify,
)
{
	const isDoc = isDocument(data);

	if (isDoc)
	{
		opts = getOptionsFromDocument(data, opts);
	}

	opts = defaultOptionsStringify(opts);

	if (isDoc)
	{
		normalizeDocument(data, opts);

		return data.toString(opts)
	}

	return stringify(data, opts)
}

/**
 * Parses Stable Diffusion wildcards source to a YAML object.
 *
 * @returns - If `Contents` extends `ParsedNode`, returns a parsed `Document.Parsed` with the specified `Contents` and `Strict`.
 *            Otherwise, returns a parsed `Document` with the specified `Contents` and `Strict`.
 *
 * @throws - Throws a `SyntaxError` if the YAML data is invalid according to the `validWildcardsYamlData` function.
 *
 * @remarks
 * This function parses the given `source` string or Uint8Array to a YAML object.
 * It uses the `parseDocument` function from the `yaml` library with `keepSourceTokens: true` option.
 * Then, it validates the parsed data using the `validWildcardsYamlData` function.
 * Finally, it returns the parsed data.
 */
export function parseWildcardsYaml<Contents extends YAMLMap = IWildcardsYAMLMapRoot, Strict extends boolean = true>(source: IParseWildcardsYamlInputSource,
	opts?: IOptionsParseDocument,
): Contents extends ParsedNode
	? IWildcardsYAMLDocumentParsed<Contents, Strict>
	: IWildcardsYAMLDocument<Contents, Strict>
{
	opts = defaultOptionsParseDocument(opts);

	if (opts.allowEmptyDocument)
	{
		source ??= '';
	}

	let data = parseDocument<Contents, Strict>(source.toString(), opts);

	// expand keys with forward slashes into nested maps
	if (opts.expandForwardSlashKeys)
	{
		_expandForwardSlashKeys(data as Document);
	}

	validWildcardsYamlData(data, opts);

	return data as any
}

export default parseWildcardsYaml

