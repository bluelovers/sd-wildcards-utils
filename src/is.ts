import { isDocument, YAMLMap, Document, isMap } from 'yaml';
import { IWildcardsYAMLDocument, IWildcardsYAMLMapRoot, IWildcardsYAMLPairValue, IWildcardsYAMLScalar } from './types';

export function isWildcardsYAMLDocument<T extends IWildcardsYAMLMapRoot>(node: IWildcardsYAMLDocument<T, true> | Document<T, true>): node is IWildcardsYAMLDocument<T, true>
export function isWildcardsYAMLDocument<T extends IWildcardsYAMLDocument | Document>(doc: any): doc is IWildcardsYAMLDocument
export function isWildcardsYAMLDocument<T extends YAMLMap = IWildcardsYAMLMapRoot>(node: any): node is IWildcardsYAMLDocument<T, true>
export function isWildcardsYAMLDocument(doc: any)
{
	return isDocument(doc)
}

export function isWildcardsYAMLDocumentAndContentsIsMap(doc: any): doc is IWildcardsYAMLDocument
{
	return isDocument(doc) && isMap(doc.contents)
}

export function isWildcardsYAMLMap<K extends IWildcardsYAMLScalar, V extends IWildcardsYAMLPairValue>(doc: IWildcardsYAMLMapRoot<K, V> | YAMLMap.Parsed<K, V> | YAMLMap<K, V>): doc is IWildcardsYAMLMapRoot<K, V>
export function isWildcardsYAMLMap<K extends IWildcardsYAMLScalar = IWildcardsYAMLScalar, V extends IWildcardsYAMLPairValue = IWildcardsYAMLPairValue>(doc: any): doc is IWildcardsYAMLMapRoot<K, V>
export function isWildcardsYAMLMap(doc: any)
{
	return isMap(doc)
}
