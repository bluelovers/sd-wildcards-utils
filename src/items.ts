import { array_unique_overwrite, defaultChecker } from 'array-hyper-unique';
import type { Alias, Document, Node, Pair, Scalar, visitor, visitorFn, YAMLMap, YAMLSeq } from 'yaml';
import { isScalar, visit } from 'yaml';
import { IOptionsParseDocument } from './options';

export type IWildcardsYAMLSeq = YAMLSeq<Scalar>;
export interface IWildcardsYAMLDocument<Contents extends YAMLMap = YAMLMap.Parsed, Strict extends boolean = true> extends Omit<Document<Contents, Strict>, 'options' | 'contents'>
{
	options: Document["options"] & IOptionsParseDocument;
	contents: Strict extends true ? Contents | null : Contents;
}

export type IWildcardsYAMLDocumentParsed<Contents extends YAMLMap = YAMLMap.Parsed, Strict extends boolean = true> =
	IWildcardsYAMLDocument<Contents, Strict>
	& Pick<Document.Parsed, 'directives' | 'range'>;

export type IOptionsVisitor = visitorFn<unknown> | {
	Alias?: visitorFn<Alias>;
	Collection?: visitorFn<YAMLMap | IWildcardsYAMLSeq>;
	Map?: visitorFn<YAMLMap>;
	Node?: visitorFn<Alias | Scalar | YAMLMap | IWildcardsYAMLSeq>;
	Pair?: visitorFn<Pair>;
	Scalar?: visitorFn<Scalar>;
	Seq?: visitorFn<IWildcardsYAMLSeq>;
	Value?: visitorFn<Scalar | YAMLMap | IWildcardsYAMLSeq>;
}

export function visitWildcardsYAML(node: Node | Document | null, visitorOptions: IOptionsVisitor)
{
	return visit(node, visitorOptions as visitor)
}

export function defaultCheckerIgnoreCase(a: unknown, b: unknown)
{
	if (typeof a === 'string' && typeof b === 'string')
	{
		a = a.toLowerCase();
		b = b.toLowerCase();
	}

	return defaultChecker(a, b)
}

export function uniqueSeqItemsChecker(a: Node, b: Node)
{
	if (isScalar(a) && isScalar(b))
	{
		return defaultCheckerIgnoreCase(a.value, b.value)
	}
	return defaultCheckerIgnoreCase(a, b)
}

export function uniqueSeqItems<T extends Node>(items: (T | unknown)[])
{
	return array_unique_overwrite(items, {
		// @ts-ignore
		checker: uniqueSeqItemsChecker,
	}) as T[];
}
