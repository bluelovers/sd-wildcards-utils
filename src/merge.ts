import { Document, isDocument, isMap, isSeq, YAMLMap } from 'yaml';

import {
	IOptionsMergeWilcardsYAMLDocumentJsonBy,
	IRecordWildcards,
	IWildcardsYAMLDocument,
	IWildcardsYAMLMapRoot, IWildcardsYAMLPair,
} from './types';
import { deepFindSingleRootAt } from './items';

export function mergeWildcardsYAMLDocumentRoots<T extends Pick<Document<YAMLMap>, 'contents'>>(ls: [T, ...any[]])
{
	return ls.reduce(_mergeWildcardsYAMLDocumentRootsCore) as T
}

export function _mergeWildcardsYAMLDocumentRootsCore<T extends Pick<Document<YAMLMap>, 'contents'>>(a: T, b: any)
{
	// @ts-ignore
	a.contents.items.push(...b.contents.items);

	return a
}

/**
 * @example
 * import { deepmergeAll } from 'deepmerge-plus';
 *
 * mergeWildcardsYAMLDocumentJsonBy(ls, {
 * 	deepmerge: deepmergeAll,
 * })
 */
export function mergeWildcardsYAMLDocumentJsonBy<T extends Document | unknown, R = IRecordWildcards>(ls: T[], opts: IOptionsMergeWilcardsYAMLDocumentJsonBy): R
{
	return opts.deepmerge(ls.map(_toJSON)) as any
}

export function _toJSON<T extends Document | unknown, R = IRecordWildcards>(v: T): R
{
	return isDocument(v) ? v.toJSON() : v
}

/**
 * Merges a single root YAMLMap or Document with a list of YAMLMap or Document.
 * The function only merges the root nodes of the provided YAML structures.
 *
 * @throws {TypeError} - If the merge target is not a YAMLMap or Document.
 * @throws {TypeError} - If the current node is not a YAMLMap.
 * @throws {TypeError} - If the current node does not support deep merge.
 */
export function mergeFindSingleRoots<T extends IWildcardsYAMLMapRoot | IWildcardsYAMLDocument>(doc: T, list: NoInfer<T>[] | NoInfer<T>): T
{
	if (!isDocument(doc) && !isMap(doc))
	{
		throw TypeError(`The merge target should be a YAMLMap or Document. doc: ${doc}`)
	}

	list = [list].flat() as NoInfer<T>[];

	for (let node of list)
	{
		let result = deepFindSingleRootAt(node);

		if (result)
		{
			let current = doc.getIn(result.paths) as IWildcardsYAMLMapRoot;

			if (current)
			{
				if (!isMap(current))
				{
					throw TypeError(`Only YAMLMap can be merged. node: ${current}`)
				}

				result.value.items
					// @ts-ignore
					.forEach((p: IWildcardsYAMLPair) => {
						let key = p.key.value;
						let sub = current.get(key);

						if (sub)
						{
							if (isSeq(sub) && isSeq(p.value))
							{
								sub.items.push(...p.value.items);
							}
							else if (isMap(sub) && isMap(p.value))
							{
								p.value.items.forEach(pair =>
								{
									sub.add(pair, false);
								})
							}
							else
							{
								delete sub.srcToken
								delete p.value.srcToken
								console.dir(result.paths.concat(key))
								console.dir(sub, {
									depth: 5,
								})
								console.dir(p.value, {
									depth: 5,
								})

								throw TypeError(`Current does not support deep merge. paths: [${result.paths.concat(key)}], a: ${sub}, b: ${p.value}`)
							}
						}
						else
						{
							current.items.push(p)
						}
					})
				;
			}
			else
			{
				doc.setIn(result.paths, result.value)
			}
		}
		else
		{
			throw TypeError(`Only YAMLMap can be merged. node: ${node}`)
		}
	}

	return doc
}
