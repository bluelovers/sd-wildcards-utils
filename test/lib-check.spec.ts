import { checkAllSelfLinkWildcardsExists } from "../src/check";
import { isSafeKey, parseWildcardsYaml, stringifyWildcardsYamlData } from '../src/index';

beforeAll(async () =>
{

});

describe(`valid`, () =>
{

	describe(`isSafeKey:true`, () =>
	{
		test.each([
			`dr._slump`,
		])(`%j`, (input) =>
		{
			let actual = isSafeKey(input);

			expect(actual).toBeTruthy();
		});
	})

	describe(`isSafeKey:false`, () =>
	{

		test.each([
			`_xxx`,
			`xxx_`,

			`-xxx`,
			`xxx-`,

			`+xxx`,
			`xxx+`,

			`@xxx`,
			`xxx@`,

			` xxx`,
			`xxx `,

			`!xxx`,
			`xxx!`,

			`~xxx`,
			`xxx~`,

			`:xxx`,
			`xxx:`,

			`/xxx`,
			`xxx/`,

			`\/xxx`,
			`xxx\/`,

			`\\xxx`,
			`xxx\\`,

			`x*x`,
			`xx__x`,

			`x#x`,
			`x x`,

			`.xxx`,
			`xxx.`,

			`'xxx`,
			`xxx'`,

			`"xxx`,
			`xxx"`,

			`x/x\\xx`,

			`x./x`,
			`x/.x`,

			`x-/x`,
			`x/-x`,

			`x_/x`,
			`x/_x`,

			`x..x`,
			//`x._x`,
			`x.-x`,
			`x_-x`,
			`x--x`,

		])(`%j`, (input) =>
		{

			let actual = isSafeKey(input);

			expect(actual).toBeFalsy();

		});

	});

});

describe(`checkAllSelfLinkWildcardsExists`, () =>
{

	/**
	 * @fixme support forward slashes `/` in keys, make `cmfy/eye_color_any` same as `cmfy:eye_color_any`
	 */
	describe(`forward slashes / in keys`, () =>
	{

		test(`cmfy/eye_color_any`, () =>
		{

			const source = `
cmfy/eye_color_classic:
  - 11
cmfy/eye_color_stylized:
  - 2
cmfy/eye_color_fantasy:
  - 3

cmfy/eye_color_any:
  - __cmfy/eye_color_classic__
  - __cmfy/eye_color_stylized__
  - __cmfy/eye_color_fantasy__
`;

			_checkAllSelfLinkWildcardsExists(source);

		});

		test(`"cmfy/eye_color_any"`, () =>
		{

			const source = `
"cmfy/eye_color_classic":
  - 1
"cmfy/eye_color_stylized":
  - 2
cmfy/eye_color_fantasy:
  - 3

"cmfy/eye_color_any":
  - __cmfy/eye_color_classic__
  - __cmfy/eye_color_stylized__
  - __cmfy/eye_color_fantasy__
`;

			_checkAllSelfLinkWildcardsExists(source);

		});

		test(`cmfy:eye_color_any`, () =>
		{

			const source = `
"cmfy/eye_color_classic":
  - 1
cmfy/eye_color_stylized:
  - 2
cmfy:
  eye_color_fantasy:
    - 3

"cmfy/eye_color_any":
  - __cmfy/eye_color_classic__
  - __cmfy/eye_color_stylized__
  - __cmfy/eye_color_fantasy__
`;

			_checkAllSelfLinkWildcardsExists(source);

		});

	})

});

describe(`jest`, () =>
{
	test(`expectToHavePropertyWithEmptyArray`, () =>
	{

		let actual = {
			errors: [],
		};

		expectToHavePropertyWithEmptyArray(actual, 'errors');

		expect(() =>
		{

			actual.errors = [''];

			expectToHavePropertyWithEmptyArray(actual, 'errors');

		}).toThrow();

	})
})

function _checkAllSelfLinkWildcardsExists(source: string)
{
	let yaml = parseWildcardsYaml(source, {
		allowMultiRoot: true,
	});

	let actual = checkAllSelfLinkWildcardsExists(yaml, {
		report: true,
	});

	let output = stringifyWildcardsYamlData(yaml);

	expect(actual).toMatchSnapshot();
	expect(output).toMatchSnapshot();

	/**
	 * @fixme actual.errors should be as empty array
	 * @fixme actual.listHasExists shoulf have length 3
	 */
	if (false)
	{
		expectToHavePropertyWithEmptyArray(actual, 'errors');

		expect(actual).toHaveProperty('listHasExists', [
			"cmfy/eye_color_classic",
			"cmfy/eye_color_stylized",
			"cmfy/eye_color_fantasy",
		])
	}

}

/**
 * Checks if an object has a property that is an empty array.
 * Used as a Jest matcher function to verify that an object has a property with an empty array value.
 * Throws an error if the property contains any elements.
 *
 * @param actual - The object to check
 * @param propertyPath - The property path to check, as a string or array
 */
function expectToHavePropertyWithEmptyArray<T extends any>(actual: T, propertyPath: string | readonly any[])
{
	expect(actual).not.toHaveProperty(propertyPath, expect.arrayContaining([expect.anything()]))
}
