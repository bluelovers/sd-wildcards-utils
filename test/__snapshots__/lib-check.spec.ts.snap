// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`checkAllSelfLinkWildcardsExists forward slashes / in keys "cmfy/eye_color_any" 1`] = `
{
  "errors": [],
  "ignoreList": [],
  "listHasExists": [
    "cmfy/eye_color_classic",
    "cmfy/eye_color_stylized",
    "cmfy/eye_color_fantasy",
  ],
  "listHasExistsWildcards": [],
  "obj": {
    "cmfy": {
      "eye_color_any": [
        "__cmfy/eye_color_classic__",
        "__cmfy/eye_color_stylized__",
        "__cmfy/eye_color_fantasy__",
      ],
      "eye_color_classic": [
        1,
      ],
      "eye_color_fantasy": [
        3,
      ],
      "eye_color_stylized": [
        2,
      ],
    },
  },
}
`;

exports[`checkAllSelfLinkWildcardsExists forward slashes / in keys "cmfy/eye_color_any" 2`] = `
"cmfy:
  eye_color_classic:
    - 1
  eye_color_stylized:
    - 2
  eye_color_fantasy:
    - 3

  eye_color_any:
    - __cmfy/eye_color_classic__
    - __cmfy/eye_color_stylized__
    - __cmfy/eye_color_fantasy__
"
`;

exports[`checkAllSelfLinkWildcardsExists forward slashes / in keys cmfy/eye_color_any 1`] = `
{
  "errors": [],
  "ignoreList": [],
  "listHasExists": [
    "cmfy/eye_color_classic",
    "cmfy/eye_color_stylized",
    "cmfy/eye_color_fantasy",
  ],
  "listHasExistsWildcards": [],
  "obj": {
    "cmfy": {
      "eye_color_any": [
        "__cmfy/eye_color_classic__",
        "__cmfy/eye_color_stylized__",
        "__cmfy/eye_color_fantasy__",
      ],
      "eye_color_classic": [
        11,
      ],
      "eye_color_fantasy": [
        3,
      ],
      "eye_color_stylized": [
        2,
      ],
    },
  },
}
`;

exports[`checkAllSelfLinkWildcardsExists forward slashes / in keys cmfy/eye_color_any 2`] = `
"cmfy:
  eye_color_classic:
    - 11
  eye_color_stylized:
    - 2
  eye_color_fantasy:
    - 3

  eye_color_any:
    - __cmfy/eye_color_classic__
    - __cmfy/eye_color_stylized__
    - __cmfy/eye_color_fantasy__
"
`;

exports[`checkAllSelfLinkWildcardsExists forward slashes / in keys cmfy:eye_color_any 1`] = `
{
  "errors": [],
  "ignoreList": [],
  "listHasExists": [
    "cmfy/eye_color_classic",
    "cmfy/eye_color_stylized",
    "cmfy/eye_color_fantasy",
  ],
  "listHasExistsWildcards": [],
  "obj": {
    "cmfy": {
      "eye_color_any": [
        "__cmfy/eye_color_classic__",
        "__cmfy/eye_color_stylized__",
        "__cmfy/eye_color_fantasy__",
      ],
      "eye_color_classic": [
        1,
      ],
      "eye_color_fantasy": [
        3,
      ],
      "eye_color_stylized": [
        2,
      ],
    },
  },
}
`;

exports[`checkAllSelfLinkWildcardsExists forward slashes / in keys cmfy:eye_color_any 2`] = `
"cmfy:
  eye_color_fantasy:
    - 3
  eye_color_classic:
    - 1
  eye_color_stylized:
    - 2

  eye_color_any:
    - __cmfy/eye_color_classic__
    - __cmfy/eye_color_stylized__
    - __cmfy/eye_color_fantasy__
"
`;
